### PR TITLE
executor, types: set correct collate in newValueExpr

### LINKED ISSUE #16279: only_full_group_by is checked wrongly
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
``` sql
create table s(a int);
select count(a) , date_format(a, '%Y-%m-%d') from s group by date_format(a, '%Y-%m-%d');
```

### 2. What did you expect to see?
``` sql
mysql> select count(a) , date_format(a, '%Y-%m-%d') from s group by date_format(a, '%Y-%m-%d');
Empty set (0.00 sec)
```

### 3. What did you see instead?
``` sql
tidb> select count(a) , date_format(a, '%Y-%m-%d') from s group by date_format(a, '%Y-%m-%d');
ERROR 1055 (42000): Expression #2 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'test.s.a' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by
```

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
`496cb870d50922ac0801307bfcfc229e1acb1009` on master

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close #16279  

Problem Summary:

### What is changed and how it works?

What's Changed:
Make the of `ValueExpr.Datum.Collate` equals to `ValueExpr.Type.Collate` in newValueExpr .

How it Works:
reflect.DeepEqual failed because of the collation of the `groupByItems` are set to `ValueExpr.FieldType.Collate`, but the collation of `ValueExpr.Datum.Collate` does not equal to `ValueExpr.FieldType.Collate`.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

N/A

### CODE REVIEW

File: types/parser_driver/value_expr.go [hunk 1]
```diff
@@ -173,7 +173,8 @@ func newValueExpr(value interface{}, charset string, collate string) ast.ValueEx
 		return ve
 	}
 	ve := &ValueExpr{}
-	ve.SetValue(value)
+	ve.Type = types.FieldType{Charset: charset, Collate: collate}
```
Comment (wjhuang2016) [inline types/parser_driver/value_expr.go:176]: The charset and collate here is used for string. Besides, the newValueExpr in parser also needs to change.

Comment (XuHuaiyu) [reply types/parser_driver/value_expr.go:176]: We do not need to check string/enum/set here.
1. Charset and collate are used for string, set, enum.
2. ve.Datum.SetValue only set the specified charset and collate for these 3 types.
3. types.DefaultTypeForValue will use default charset and collate if the type is not string/enum/set.

Comment (XuHuaiyu) [reply types/parser_driver/value_expr.go]: And we do not need to change the implementation for parser now.
The interface of Datum in parser is not the same as tidb.

**Comment by XuHuaiyu** [comment]
PTAL @qw4990

**Pull request merged**